src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs

```diff
@@ -673,19 +683,56 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
+            if (builder == null)
             {
-                left = VisitExpression(ref builder, node.Left);
+                left = VisitExpression(ref builder, left);
             }
             else
             {
                 // if the right-hand-side has await, spill the left
                 var leftBuilder = new BoundSpillSequenceBuilder();
-                left = VisitExpression(ref leftBuilder, node.Left);
-                if (left.Kind != BoundKind.Local)
+
+                switch (left.Kind)
                 {
-                    left = Spill(leftBuilder, left, RefKind.Ref);
+                    case BoundKind.Local:
+                    case BoundKind.Parameter:
+                        // locals and parameters are directly assignable, LHS is not on the stack so nothing to spill
+                        break;
+
+                    case BoundKind.FieldAccess:
+                        var field = (BoundFieldAccess)left;
+                        // static fields are directly assignable, LHS is not on the stack, nothing to spill
+                        if (field.FieldSymbol.IsStatic) break;
+
+                        // instance fields are directly assignable, but receiver is pushed, so need to spill that.
+                        var receiver = VisitExpression(ref leftBuilder, field.ReceiverOpt);
+                        receiver = Spill(builder, receiver, field.FieldSymbol.ContainingType.IsValueType ? RefKind.Ref : RefKind.None);
+                        left = field.Update(receiver, field.FieldSymbol, field.ConstantValueOpt, field.ResultKind, field.Type);
+                        break;
+
+                    case BoundKind.ArrayAccess:
+                        var arrayAccess = (BoundArrayAccess)left;
+                        // array and indices are pushed on stack so need to spill that
+                        var expression = VisitExpression(ref leftBuilder, arrayAccess.Expression);
+                        expression = Spill(builder, expression, RefKind.None);
+                        var indices = this.VisitExpressionList(ref builder, arrayAccess.Indices, forceSpill: true);
+                        left = arrayAccess.Update(expression, indices, arrayAccess.Type);
+                        break;
+
+                    default:
+                        // must be something indirectly assignable, just visit and spill as an ordinary Ref  (not a RefReadOnly!!)
+                        //
+                        // NOTE: in some cases this will result in spiller producing an error.
+                        //       For example if the LHS is a ref-returning method like
+                        //
+                        //       obj.RefReturning(a, b, c) = await Something();
+                        //
+                        //       the spiller would eventually have to spill the evaluation result of "refReturning" call as an ordinary Ref, 
+                        //       which it can't.
```

Comment:
apparently not. I have a general bug to follow up with async spilling test coverage [https://github.com/dotnet/roslyn/issues/20890](https://github.com/dotnet/roslyn/issues/20890) 
For this case I have added a test to be sure - CannotUseAwaitExpressionToAssignRefReturing.

---
In reply to: [127541951](https://github.com/dotnet/roslyn/pull/20883#discussion_r127541951) [](ancestors = 127541951)